src/visidentifier/visidentifier.cpp

```diff
@@ -0,0 +1,403 @@
+
+    } catch (const AosException& e) {
+        LOG_ERR() << e.what();
+    }
+}
+
+void VISIdentifier::HandleConnection()
+{
+    do {
+        try {
+            mWsClientPtr->Connect();
+
+            Subscribe(
+                cSubjectsVISPath, std::bind(&VISIdentifier::HandleSubjectsSubscription, this, std::placeholders::_1));
+
+            mSystemId.Clear();
+            mUnitModel.Clear();
+            mSubjects.Clear();
+
+            mWSClientIsConnected.set();
+
+            // block on Wait
+            const auto wsClientEvent = mWsClientPtr->WaitForEvent();
+
+            if (wsClientEvent.mCode == WSClientEvent::EventEnum::CLOSED) {
+                LOG_INF() << "WS Client connection has been closed. Stopping Vis Identifier Handle Connection thread";
+
+                return;
+            }
+
+            mWSClientIsConnected.reset();
+            mWsClientPtr->Disconnect();
+
+        } catch (const WSException& e) {
+            LOG_ERR() << "WSException has been caught: message = " << e.message().c_str();
+
+            mWSClientIsConnected.reset();
+            mWsClientPtr->Disconnect();
+        } catch (const Poco::Exception& e) {
+            LOG_ERR() << "Poco::Exception has been caught: message = " << e.message().c_str();
+        } catch (...) {
+            LOG_ERR() << "Unknown exception caught.";
+        }
+    } while (!mStopHandleSubjectsChangedThread.tryWait(cWSClientReconnectMilliseconds));
+}
+
+aos::Error VISIdentifier::HandleSubjectsSubscription(Poco::Dynamic::Var value)
+{
+    try {
+        aos::StaticArray<aos::StaticString<aos::cSubjectIDLen>, aos::cMaxSubjectIDSize> newSubjects;
+
+        const auto responseSubjects = GetValueArrayByPath(value, cSubjectsVISPath);
+
+        for (const auto& subject : responseSubjects) {
+            if (auto err = newSubjects.PushBack(subject.c_str()); !err.IsNone()) {
+                return err;
+            }
+        }
+
+        std::lock_guard lock(mMutex);
+
+        if (mSubjects != newSubjects) {
+            mSubjects = std::move(newSubjects);
+            mSubjectsObserverPtr->SubjectsChanged(mSubjects);
+        }
+    } catch (const Poco::Exception& e) {
+        LOG_ERR() << e.message().c_str();
+
+        return aos::ErrorEnum::eFailed;
+    }
+
+    return aos::ErrorEnum::eNone;
+}
+
+std::string VISIdentifier::SendGetRequest(const std::string& path)
+{
+    const auto       requestId = mWsClientPtr->GenerateRequestID();
+    const VISMessage getMessage(VISActionEnum::eGet, requestId, path);
+
+    mWSClientIsConnected.wait();
+
+    const auto response = mWsClientPtr->SendRequest(requestId, getMessage.ToByteArray());
+
+    return {response.cbegin(), response.cend()};
+}
+
+void VISIdentifier::SendUnsubscribeAllRequest()
+{
+    try {
+        const VISMessage request(VISActionEnum::eUnsubscribeAll, mWsClientPtr->GenerateRequestID(), "");
+
+        mWsClientPtr->SendMessage(request.ToByteArray());
```

Comment:
i think in this case we can have race condition, so when we try to send message in parallel we could disconnect from websocket

Reply:
PocoWSClient::SendMessage is thread safe
If case of error: connection failed, sendFrame failed - WSException is thrown and caller code handles it